### ISSUE TITLE
Tensor.variance(dim:int) bug: sometimes it gives a negative variance

### ISSUE THREAD

**Issue by nhirschey** [DiffSharp/DiffSharp#336]
There appears to be a bug in `Tensor.variance(dim:int, ?keepDim:bool, ?unbiased:bool)` such that it can give a negative variance.

```fsharp
#r "nuget: DiffSharp-cuda-windows, 1.0.0-preview-783523654"

open DiffSharp

dsharp.config(dtype=Dtype.Float32,device=Device.CPU,backend=Backend.Torch)

let test =
    dsharp.tensor([[996.52, -8.02],
        [996.57, -8.41],
        [996.53, -8.51],
        [996.51, -8.31],
        [996.51, -8.27],
        [996.5, -8.05],
        [996.5, -7.62],
        [996.5, -7.62]],dtype=Float32)

// Here we get a negative variance for the first column of data.
test.variance(0)
(** output:
val it : Tensor = tensor([-0.0714286, 0.115278])
*)

// Here we get correct variance
test.[0..,0].variance()
(** output:
val it : Tensor = tensor(0.000564476) {Item = ?;
                                       backend = Torch;
                                       depth = 0;
                                       derivative = ?;
                                       derivativeDeep = ?;
                                       device = Device (CPU, -1);
                                       deviceType = CPU;
                                       dim = 0;
                                       dtype = Float32;
                                       fanout = ?;
                                       nelement = 1;
                                       nestingTag = ?;
                                       parentOp = ?;
                                       primal = ...;
                                       primalDeep = ...;
                                       shape = [||];}
*)


```

And a sanity check:
```fsharp
let correctVariance =
    Array.init test.shape.[0] (fun i ->  test.[i,0].toSingle())
    |> fun xs ->
        let mu = xs |> Array.average
        let n = xs.Length - 1 |> single
        xs |> Array.sumBy(fun x -> (1f/n)*(x-mu)**2.0f)

(** output:
val correctVariance : float32 = 0.000564472517f
*)
```

**Comment by dsyme** [comment]
Note sum/variance need to be reimplemented https://github.com/DiffSharp/DiffSharp/issues/279

**Comment by gbaydin** [comment]
@nhirschey thank you for catching and isolating this!

I think this is happening because the `.variance()` and `.variance(dim:int)` versions used different implementations, referred to as "two-pass" and "naive" in this page https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

The naive one used in `.variance(dim:int)` is not the best one and there was actually a comment in the source code to improve this if/when we hit problems. I think this is the moment to improve it thanks to you.

I'm now looking at the fix in #357 that you submitted.

**Comment by gbaydin** [comment]
Fixed by #357

**Comment by nhirschey** [comment]
Yes, I saw the code comment but I was still surprised it turned out to be the naive vs. two-pass difference rearing its head in a standard dataset. It prevented me from doing t.variance(0) on a dataset from Francois Chollet's Deep Learning with Python book. 

The precision unit test that I added will be a safeguard for when this is rewritten following  #279. 

And thank you for your work on this wonderful library.
